diff --git a/ttnn/cpp/ttnn/operations/data_movement/gather/device/kernels/dataflow/gather_reader_single_row_multi_core.cpp b/ttnn/cpp/ttnn/operations/data_movement/gather/device/kernels/dataflow/gather_reader_single_row_multi_core.cpp
index b78625cd1036..e4111d9c2d3c 100644
--- a/ttnn/cpp/ttnn/operations/data_movement/gather/device/kernels/dataflow/gather_reader_single_row_multi_core.cpp
+++ b/ttnn/cpp/ttnn/operations/data_movement/gather/device/kernels/dataflow/gather_reader_single_row_multi_core.cpp
@@ -81,10 +81,12 @@ void kernel_main() {
     constexpr uint32_t output_tensor_data_format_size =
         get_tile_size(output_tensor_cb_index) / get_tile_hw(input_tensor_cb_index);
 
-    uint32_t current_index_tile_id = core_id;
-
     for (uint32_t h = 0; h < Ht; h++) {
         for (uint32_t core_loop = 0; core_loop < core_loop_count; core_loop++) {
+            const uint32_t current_index_tile_id = core_id + core_loop * total_number_of_cores;
+            if (current_index_tile_id >= Wt_index) {
+                break;
+            }
             // Read index data
             cb_reserve_back(input_index_tensor_cb_index, one_tile);
 
@@ -150,7 +152,6 @@ void kernel_main() {
             }  // wi loop
             cb_push_back(output_tensor_cb_index, one_tile);
             cb_pop_front(input_index_tensor_cb_index, one_tile);
-            current_index_tile_id += total_number_of_cores;
         }  // core_loop_count loop
     }  // h loop
 }
diff --git a/ttnn/cpp/ttnn/operations/data_movement/gather/device/kernels/dataflow/gather_writer_single_row_multi_core.cpp b/ttnn/cpp/ttnn/operations/data_movement/gather/device/kernels/dataflow/gather_writer_single_row_multi_core.cpp
index 1cf6d926edf3..3efa887acdf2 100644
--- a/ttnn/cpp/ttnn/operations/data_movement/gather/device/kernels/dataflow/gather_writer_single_row_multi_core.cpp
+++ b/ttnn/cpp/ttnn/operations/data_movement/gather/device/kernels/dataflow/gather_writer_single_row_multi_core.cpp
@@ -46,10 +46,12 @@ void kernel_main() {
     const auto output_tensor_dram =
         TensorAccessor(output_tensor_args, output_tensor_buffer_addr, output_tensor_tile_size_bytes);
 
-    uint32_t current_index_tile_id = core_id;
-
     for (uint32_t h = 0; h < Ht; h++) {
         for (uint32_t core_loop = 0; core_loop < core_loop_count; core_loop++) {
+            const uint32_t current_index_tile_id = core_id + core_loop * total_number_of_cores;
+            if (current_index_tile_id >= Wt_index) {
+                break;
+            }
             // Read input data
             for (uint32_t w = 0; w < Wt_input; w++) {
                 cb_reserve_back(input_tensor_cb_index, one_tile);
@@ -67,7 +69,6 @@ void kernel_main() {
             noc_async_write_barrier();
             cb_pop_front(output_tensor_cb_index, one_tile);
 
-            current_index_tile_id += total_number_of_cores;
         }  // core_loop_count loop
     }  // h loop
 }
